### PR TITLE
Proper constant folding in the abstract interpreter

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
This PR is a much simpler version of https://github.com/p4lang/p4c/pull/2915/files.
@vladyslav-dubina : if there is a test case where the compiler crashes with these changes please post it as a comment.

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
There are no new test cases because this code is only exercised when unrolling loops, and for our standard p4test tests loop unrolling is disabled.

**Comment by mbudiu-vmw** [comment]
For example, the following command used to cause a bug: `./p4test --loopsUnroll ../testdata/p4_16_samples/issue361-bmv2.p4`.

**Comment by mbudiu-vmw** [comment]
@VolodymyrPeschanenko @VolodymyrPeschanenkoIntel (which one is the right id?) - can you please review this?

**Comment by mbudiu-vmw** [comment]
BTW: @VolodymyrPeschanenko you actually don't have reviewer privileges, but I think you should. Would you be interested to receive them?

**Comment by mbudiu-vmw** [comment]
@vladyslav-dubina I can see 3 failed tests with the cast issue you mentioned, I will take a look.

**Comment by VolodymyrPeschanenkoLitSoft** [comment]
> BTW: @VolodymyrPeschanenko you actually don't have reviewer privileges, but I think you should. Would you be interested to receive them?

@mbudiu-vmw Thanks for your trust. I am ready to help with the reviews. As you know I am always in touch, but, please, notice the time difference with Ukraine.

**Comment by jnfoster** [comment]
Added. Welcome aboard!

**Review by VolodymyrPeschanenkoLitSoft — APPROVED** [review]
These changes look fine.

**Comment by VolodymyrPeschanenkoLitSoft** [comment]
> Added. Welcome aboard!

Thanks.

**Review by VolodymyrPeschanenkoLitSoft — CHANGES_REQUESTED** [review]
Sorry, I need to check it again.

**Comment by VolodymyrPeschanenkoLitSoft** [comment]
> ./p4test --loopsUnroll ../testdata/p4_16_samples/issue361-bmv2.p4

Speaking true, it passed on my laptop, but these changes look fine. Anyway, I think that these changes should be applied [here](https://github.com/mbudiu-vmw/p4c-clone/blob/7fbc2a4dfb4427b9c1122f66ecca4532d1612d2b/midend/interpreter.cpp#L723) and  [here](https://github.com/mbudiu-vmw/p4c-clone/blob/7fbc2a4dfb4427b9c1122f66ecca4532d1612d2b/midend/interpreter.cpp#L732) too.

File: midend/interpreter.cpp
Comment (VolodymyrPeschanenkoLitSoft) [inline midend/interpreter.cpp:869]: I think that we need to do the same thing for other cases.

Comment (mbudiu-vmw) [reply midend/interpreter.cpp:869]: The reason we don't need to set the type in the typemap in the other cases is because the constant folding in general does not need to know types - but it does need to know if for casts.

**Comment by mbudiu-vmw** [comment]
@VolodymyrPeschanenko you need to give an official "Approve" if this PR is to be merged (review changes/approve)

**Review by VolodymyrPeschanenkoLitSoft — APPROVED** [review]

**Comment by mbudiu-vmw** [comment]
You have too many accounts, I don't know which one has been granted reviewer permissions, but this it's not the one you have used to approve...

**Review by VolodymyrPeschanenko — APPROVED** [review]
Let's try with this account.

**Comment by mbudiu-vmw** [comment]
As long as it says "review required" it didn't work.
@jnfoster which account did you approve?

**Review by VolodymyrPeschanenkoIntel — APPROVED** [review]
This is last my account on github.

**Comment by mbudiu-vmw** [comment]
I think @jnfoster mistyped something - this didn't work either.

**Comment by VolodymyrPeschanenko** [comment]
@mbudiu-vmw, It looks like I don't have write access for any of my accounts in github. @VolodymyrPeschanenkoLitSoft is my main account.

**Comment by jnfoster** [comment]
Oops. I cut and paste the username @mbudiu-vmw wrote above, which was indeed the wrong account. Fixed now :-)

**Pull request merged**
